cppcache/integration/test/ContainsKeyOnServerExceptionTest.cpp

```diff
@@ -0,0 +1,111 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <framework/Cluster.h>
+#include <framework/TestConfig.h>
+
+#include <geode/Cache.hpp>
+#include <geode/CacheFactory.hpp>
+#include <geode/CacheTransactionManager.hpp>
+#include <geode/Region.hpp>
+#include <geode/RegionFactory.hpp>
+#include <geode/RegionShortcut.hpp>
+
+namespace {
+using apache::geode::client::Cache;
+using apache::geode::client::CacheableKey;
+using apache::geode::client::CacheableString;
+using apache::geode::client::CacheFactory;
+using apache::geode::client::CacheTransactionManager;
+using apache::geode::client::Pool;
+using apache::geode::client::Region;
+using apache::geode::client::RegionShortcut;
+
+std::string disconnectingStr("Disconnecting from the endpoint");
+
+std::string getClientLogName() {
+  std::string testSuiteName(::testing::UnitTest::GetInstance()
+                                ->current_test_info()
+                                ->test_case_name());
+  std::string testCaseName(
+      ::testing::UnitTest::GetInstance()->current_test_info()->name());
+  std::string logFileName(testSuiteName + "/" + testCaseName + "/client.log");
+  return logFileName;
+}
+
+std::shared_ptr<Cache> createCache() {
+  auto cache = CacheFactory()
+                   .set("log-level", "none")
+                   .set("log-file", getClientLogName())
+                   .create();
+  return std::make_shared<Cache>(std::move(cache));
+}
+
+void executeTestCaseHandleException() {
+  Cluster cluster{
+      LocatorCount{1}, ServerCount{3},
+      CacheXMLFiles({
+          std::string(getFrameworkString(FrameworkVariable::TestCacheXmlDir)) +
+              "/cacheserver1_fpr_transaction.xml",
+          std::string(getFrameworkString(FrameworkVariable::TestCacheXmlDir)) +
+              "/cacheserver2_fpr_transaction.xml",
+          std::string(getFrameworkString(FrameworkVariable::TestCacheXmlDir)) +
+              "/cacheserver3_fpr_transaction.xml",
+      })};
+  cluster.start([&]() {
+    cluster.getGfsh()
+        .deploy()
+        .jar(getFrameworkString(FrameworkVariable::JavaObjectJarPath))
+        .execute();
+  });
+
+  auto cache = createCache();
+  auto poolFactory = cache->getPoolManager().createFactory();
+  ServerAddress serverAddress = cluster.getServers()[0].getAddress();
+  serverAddress.port = 40401;
+  cluster.applyServer(poolFactory, serverAddress);
+  auto pool = poolFactory.create("default");
+  auto region = cache->createRegionFactory(RegionShortcut::PROXY)
+                    .setPoolName("default")
+                    .create("region");
+
+  auto transactionManager = cache->getCacheTransactionManager();
+
+  // this key will be always routed towards server[1]
+  int theKey = 7;
+  std::string theValue = "theValue";
+  try {
```

Comment:
try/catches in a test always make me a little concerned.  This probably comes down to one's testing philosophy, but I generally try to avoid them and let the errors explode.  It seems in this set of try/catches, you are only printing that an exception was caught, no changes in assertions or test expectations.

Reply:
Can you help me with this, as I have no idea how to do that with assertions?
The main goal is to get proper exception on containsKeyOnServer function.

Reply:
I'm having trouble running the test as is on my Ubuntu box, so not sure if this works.  But was maybe thinking something like

```c++
  try {
    transactionManager->begin();
    region->put(theKey, theValue);
    transactionManager->commit();
  } catch (...) {
    EXPECT_THROW(region->containsKeyOnServer(CacheableKey::create(7)), apache::geode::client::TransactionDataNodeHasDepartedException);
  }
```

Which removes at least the nested try/catch.  And asserts that the expect exception is thrown, which will produce a better message if/when the test fails.  Maybe @pdxcodemonkey or @echobravopapa have some better ideas for removin the first try/catch, but that may not be as big a deal.

Reply:
I run it with your changes and it passed with the fix, and without the fix it failed. So the test is ok and it works.

Reply:
This seems good, though I have one concern and I'm not sure how to resolve it.  Since the `EXPECT_THROW` is in the catch, that expectation will only run if the `try` fails.  The `EXPECT_FALSE` will check that region doesn't contain the key, but the test wouldn't prove the right exception happened.  If I get to machine where I can run this (or if you want to try it out before me), it would be worth testing that the correct exception is being thrown and tested by changing the `EXPECT_THROW` to the wrong exception and see the test fail.  That would give us more confidence that the test is working as intended.

Otherwise I think everything looks alright.

Reply:
Hi @moleske ,
I tested this, and if I set the wrong exception in `EXPECT_THROW`, test failed.
If I set the right exception(`TransactionDataNodeHasDepartedException`) it passed.
Thanks for a help!